miniapps/autodiff/par_example.cpp

```diff
@@ -0,0 +1,521 @@
+// Copyright (c) 2010-2021, Lawrence Livermore National Security, LLC. Produced
+// at the Lawrence Livermore National Laboratory. All Rights reserved. See files
+// LICENSE and NOTICE for details. LLNL-CODE-806117.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability visit https://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+//
+//                       MFEM AD Example  - Parallel Version
+//
+// Compile with: make par_example
+//
+// Sample runs:
+//   mpirun -np 2 paradiff -m ../data/beam-quad.mesh -pp 3.8
+//   mpirun -np 2 paradiff -m ../data/beam-tri.mesh  -pp 7.2
+//   mpirun -np 2 paradiff -m ../data/beam-hex.mesh
+//   mpirun -np 2 paradiff -m ../data/beam-tet.mesh
+//   mpirun -np 2 paradiff -m ../data/beam-wedge.mesh
+//
+// Description:  This examples solves a quasi-static nonlinear
+//               p-Laplacian problem with zero Dirichlet boundary
+//               conditions applied on all defined boundaries
+//
+//           The example demonstrates the use of nonlinear operators combined
+//           with automatic differentiation (AD). The definitions of the
+//           integrators are written in the example.hpp.  Selecting integrator=0
+//           will use the manually implemented integrator.  Selecting
+//           integrator=1,2 will utilize one of the AD integrators.
+//
+//        We recommend viewing examples 1 and 19, before playing with this
+//        example.
+
+#include "example.hpp"
+
+using namespace mfem;
+
+///Non-linear solver for the p-Laplacian problem.
+class ParNLSolverPLaplacian
+{
+public:
+   ///Constructor Input: imesh - FE mesh, finite element space,
+   /// power for the p-Laplacian, external load (source, input),
+   /// regularization parameter
+   ParNLSolverPLaplacian(MPI_Comm comm, ParMesh& imesh,
+                         ParFiniteElementSpace& ifespace,
+                         double powerp=2,
+                         Coefficient* load=nullptr,
+                         double regularizationp=1e-7)
+   {
+      lcomm = comm;
+
+      //default parameters for
+      //the Newton solver
+      newton_rtol = 1e-4;
+      newton_atol = 1e-8;
+      newton_iter = 10;
+
+      //linear solver
+      linear_rtol = 1e-7;
+      linear_atol = 1e-15;
+      linear_iter = 500;
+
+      print_level = 0;
+
+      //set the mesh
+      mesh=&imesh;
+
+      //set the fespace
+      fespace=&ifespace;
+
+      //set the parameters
+      plap_epsilon=new ConstantCoefficient(regularizationp);
```

Comment:
Recommend using std::unique_ptr instead of new/delete. new/delete may make sense for low level code, but at the high level like this, the safety seems well worth it.

Reply:
My understanding is that the MFEM team does not want to use smart pointers in the code. Otherwise I will be happy to introduce smart pointers everywhere.

Reply:
It is up to you, but I think it is OK to use them in the miniapp.

The concern is about introducing them in the core library where they will be mixed with objects managed by new/delete. (We have discussed overhauling the memory management for all objects, but that will be a really big change)

Reply:
I suspected that this was the case, and I agree with that in contexts where the overhead would matter. From what I've read, the overhead of std::unique is negligible in most cases. It involves things like:

- initializing the pointer to nullptr in the default constructor -- something that you're doing already
- if you provide a custom delete function (not the default case), there is some memory cost and cost associated with copying the function pointer. This cost is not present in the case I am recommending.
- it frees the pointer and runs the destructor (if any) on the reference object if the variable goes out of scope. It is by far the easiest method to ensure that memory is not leaked in the presence of an exception

Most of these minimal costs are really virtues in disguise. In the vast majority of cases, it's better to initialize a pointer to nullptr if you aren't initializing it with an allocated pointer initially. Deleting the pointer automatically is also a benefit that prevents memory leaks. If you don't use std::unique_ptr, you should really have try catch blocks all over the place to ensure that allocated pointers are freed in the presence of exceptions. Even if MFEM never throws exceptions, client code may. Only 10 methods in the installed header files for MFEM 4.2 are declared noexcept, to the client it appears that MFEM does throw exceptions. Note that I am advocating for std::unique_ptr here -- not std::shared_ptr.

There are follow-on impacts to MFEM's policy. When someone needs to implement a modified or specialized version of something in one of the miniapps or examples, everyone starts by cutting and pasting MFEM's code into their code with the sense that they're utilizing the best practices because they're doing it the MFEM-way. I've seen the patterns in this code in several places in the LiDO code, and when I touch that LiDO code, I generally replace the new/delete's with std::unique_ptr. The new/delete approach runs the risk of uninitialized pointers, forgotten delete's causing leaks, and exceptions causing leaks.

Reply:
I would prefer to keep the old fashion way in the examples in order to keep the focus on the AD instead of adding "new" techniques to the code. I will use smart pointers in the next example.